File: frontend/seo/ogImage.ts
/**
 * Draws the 1200x630 social card: circular photo on the left, name, role, a
 * skills line and the domain along the bottom.
 *
 * When no photo is available the text column takes the full width instead.
 */

import sharp, { type OverlayOptions } from 'sharp';
import type { SiteMetadata } from './metadata';

export const OG_WIDTH = 1200;
export const OG_HEIGHT = 630;

/** Distance every drawn element keeps from the edges. */
export const SAFE_MARGIN = 60;

export const OG_COLORS = {
  background: '#0d1f36',
  name: '#f5f7fa',
  role: '#c8d3e0',
  accent: '#2ec4c6',
  domain: '#8fa3bd',
};

const FONT_STACK = "'DejaVu Sans', 'Liberation Sans', 'Helvetica Neue', Arial, sans-serif";

const PHOTO_DIAMETER = 300;
const PHOTO_LEFT = 90;
const PHOTO_TOP = 165;
const PHOTO_RING_WIDTH = 6;

const TEXT_LEFT_WITH_PHOTO = 460;
const TEXT_LEFT_ALONE = SAFE_MARGIN;
const TEXT_RIGHT = OG_WIDTH - SAFE_MARGIN;

const NAME_MAX_SIZE = 84;
const NAME_MIN_SIZE = 44;
const ROLE_MAX_SIZE = 40;
const ROLE_MIN_SIZE = 26;
const SKILLS_SIZE = 28;
const DOMAIN_SIZE = 24;
const DOMAIN_LETTER_SPACING = 3;
const DOMAIN_BASELINE = 552;

/** Skills joined into the line under the role. */
const SKILLS_SHOWN = 3;
const SKILLS_SEPARATOR = ' · ';

/**
 * Per-character width as a fraction of the font size, approximating the metrics
 * of the sans-serif stack above. Used to shrink text that would otherwise run
 * past the safe margin.
 */
function characterWidth(char: string): number {
  if (char === ' ') return 0.32;
  if ('iljtI.,:;\'!|[]()/\\"`'.includes(char)) return 0.3;
  if ('fr'.includes(char)) return 0.38;
  if ('mw'.includes(char)) return 0.9;
  if ('MW'.includes(char)) return 0.95;
  if (char >= 'A' && char <= 'Z') return 0.68;
  if (char >= '0' && char <= '9') return 0.64;
  return 0.56;
}

export function estimateTextWidth(text: string, fontSize: number, letterSpacing = 0): number {
  const glyphs = [...text].reduce((total, char) => total + characterWidth(char), 0) * fontSize;
  return glyphs + Math.max(0, text.length - 1) * letterSpacing;
}

/** The largest size in the range at which `text` fits `maxWidth`. */
export function fitFontSize(
  text: string,
  maxWidth: number,
  maxSize: number,
  minSize: number
): number {
  for (let size = maxSize; size > minSize; size -= 1) {
    if (estimateTextWidth(text, size) <= maxWidth) return size;
  }
  return minSize;
}

/** Shortens `text` with a trailing ellipsis until it fits `maxWidth`. */
export function truncateToWidth(
  text: string,
  fontSize: number,
  maxWidth: number,
  letterSpacing = 0
): string {
  if (estimateTextWidth(text, fontSize, letterSpacing) <= maxWidth) return text;

  let clipped = text;
  while (
    clipped.length > 1 &&
    estimateTextWidth(`${clipped}\u2026`, fontSize, letterSpacing) > maxWidth
  ) {
    clipped = clipped.slice(0, -1);
  }

  return `${clipped.trimEnd()}\u2026`;
}

/** The text and size to draw so a line fills the available width without passing it. */
export function fitText(
  text: string,
  maxWidth: number,
  maxSize: number,
  minSize: number,
  letterSpacing = 0
): { text: string; fontSize: number } {
  const fontSize = fitFontSize(text, maxWidth, maxSize, minSize);
  return { text: truncateToWidth(text, fontSize, maxWidth, letterSpacing), fontSize };
}

function escapeXml(value: string): string {
  return value
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

/** Joins as many leading skills as fit the available width, or an empty string if none do. */
export function fitSkillsLine(skills: string[], maxWidth: number): string {
  for (let count = Math.min(SKILLS_SHOWN, skills.length); count > 0; count -= 1) {
    const line = skills.slice(0, count).join(SKILLS_SEPARATOR);
    if (estimateTextWidth(line, SKILLS_SIZE) <= maxWidth) return line;
  }
  return '';
}

function textElement(
  content: string,
  x: number,
  y: number,
  fontSize: number,
  fill: string,
  extra = ''
): string {
  return (
    `<text x="${x}" y="${y}" font-family="${FONT_STACK}" font-size="${fontSize}" ` +
    `fill="${fill}"${extra}>${escapeXml(content)}</text>`
  );
}

/** Builds the SVG layer drawn over the background and the photo. */
export function buildOverlaySvg(metadata: SiteMetadata, hasPhoto: boolean): string {
  const left = hasPhoto ? TEXT_LEFT_WITH_PHOTO : TEXT_LEFT_ALONE;
  const width = TEXT_RIGHT - left;

  const name = fitText(metadata.name, width, NAME_MAX_SIZE, NAME_MIN_SIZE);
  const role = fitText(metadata.jobTitle, width, ROLE_MAX_SIZE, ROLE_MIN_SIZE);
  const skillsLine = fitSkillsLine(metadata.skills, width);

  const nameBaseline = hasPhoto ? 258 : 268;
  const roleBaseline = nameBaseline + 60;
  const skillsBaseline = roleBaseline + 48;

  const parts = [
    textElement(name.text, left, nameBaseline, name.fontSize, OG_COLORS.name),
    textElement(role.text, left, roleBaseline, role.fontSize, OG_COLORS.role),
  ];

  if (skillsLine) {
    parts.push(textElement(skillsLine, left, skillsBaseline, SKILLS_SIZE, OG_COLORS.accent));
  }

  parts.push(
    textElement(
      truncateToWidth(metadata.host.toUpperCase(), DOMAIN_SIZE, width, DOMAIN_LETTER_SPACING),
      left,
      DOMAIN_BASELINE,
      DOMAIN_SIZE,
      OG_COLORS.domain,
      ` letter-spacing="${DOMAIN_LETTER_SPACING}"`
    )
  );

  if (hasPhoto) {
    const radius = PHOTO_DIAMETER / 2;
    parts.unshift(
      `<circle cx="${PHOTO_LEFT + radius}" cy="${PHOTO_TOP + radius}" ` +
        `r="${radius - PHOTO_RING_WIDTH / 2}" fill="none" ` +
        `stroke="${OG_COLORS.accent}" stroke-width="${PHOTO_RING_WIDTH}" />`
    );
  }

  return (
    `<svg xmlns="http://www.w3.org/2000/svg" width="${OG_WIDTH}" height="${OG_HEIGHT}">` +
    `${parts.join('')}</svg>`
  );
}

/** Crops the photo to a circle of the card's photo diameter. */
async function circularPhoto(photo: Buffer): Promise<Buffer> {
  const radius = PHOTO_DIAMETER / 2;
  const mask = Buffer.from(
    `<svg xmlns="http://www.w3.org/2000/svg" width="${PHOTO_DIAMETER}" height="${PHOTO_DIAMETER}">` +
      `<circle cx="${radius}" cy="${radius}" r="${radius}" fill="#ffffff" /></svg>`
  );

  return sharp(photo)
    .resize(PHOTO_DIAMETER, PHOTO_DIAMETER, { fit: 'cover', position: 'centre' })
    .composite([{ input: mask, blend: 'dest-in' }])
    .png()
    .toBuffer();
}

/**
 * Renders the card as a PNG.
 *
 * A photo that cannot be decoded is dropped and the text-only layout is used.
 */
export async function renderOgImage(metadata: SiteMetadata, photo: Buffer | null): Promise<Buffer> {
  let circle: Buffer | null = null;

  if (photo) {
    try {
      circle = await circularPhoto(photo);
    } catch {
      circle = null;
    }
  }

  const layers: OverlayOptions[] = [];
  if (circle) layers.push({ input: circle, left: PHOTO_LEFT, top: PHOTO_TOP });
  layers.push({ input: Buffer.from(buildOverlaySvg(metadata, circle !== null)), left: 0, top: 0 });

  return sharp({
    create: {
      width: OG_WIDTH,
      height: OG_HEIGHT,
      channels: 4,
      background: OG_COLORS.background,
    },
  })
    .composite(layers)
    .png({ compressionLevel: 9 })
    .toBuffer();
}

File: tools/seo/og_image.py
"""Draws the 1200x630 social card: circular photo on the left, name, role, a
skills line and the domain along the bottom.

The card is set in the site's own webfonts, read out of the built frontend.
Without a usable photo the text column takes the full width instead.
"""

from __future__ import annotations

import io
from collections.abc import Callable, Sequence
from functools import partial
from pathlib import Path

from fontTools.ttLib import TTFont
from PIL import Image, ImageDraw, ImageFont, ImageOps

WIDTH = 1200
HEIGHT = 630

#: Distance every drawn element keeps from the edges.
SAFE_MARGIN = 60

COLORS = {
    "background": "#0d1f36",
    "name": "#f5f7fa",
    "role": "#c8d3e0",
    "accent": "#2ec4c6",
    "domain": "#8fa3bd",
}

#: Site faces, relative to the directory the built frontend serves them from.
DISPLAY_FONT_FILE = "fonts/cormorant-garamond-latin.woff2"
BODY_FONT_FILE = "fonts/dm-sans-latin.woff2"

#: The photo, and the factor its mask and ring are drawn at before being scaled
#: down, which smooths their edges.
PHOTO_DIAMETER, PHOTO_LEFT, PHOTO_TOP, RING_WIDTH = 300, 90, 165, 6
SUPERSAMPLE = 4

#: Where the text column starts, with a photo beside it and without one.
TEXT_LEFT_WITH_PHOTO, TEXT_LEFT_ALONE = 460, SAFE_MARGIN
TEXT_RIGHT = WIDTH - SAFE_MARGIN

#: Largest and smallest size the name and the role are set at.
NAME_SIZES, ROLE_SIZES = (84, 44), (40, 26)
SKILLS_SIZE, DOMAIN_SIZE, DOMAIN_TRACKING = 28, 24, 3

#: Baseline of the name, with a photo beside it and without one, and the
#: baselines of the lines below it.
NAME_BASELINE_WITH_PHOTO, NAME_BASELINE_ALONE = 258, 268
ROLE_OFFSET, SKILLS_OFFSET, DOMAIN_BASELINE = 60, 108, 552

#: Skills joined into the line under the role.
SKILLS_SHOWN, SKILLS_SEPARATOR = 3, " · "

ELLIPSIS = "…"

Font = ImageFont.FreeTypeFont


def truetype_bytes(path: Path) -> bytes | None:
    """The TrueType bytes of a woff2 file, which FreeType cannot read directly."""
    try:
        font = TTFont(io.BytesIO(path.read_bytes()))
        font.flavor = None
        buffer = io.BytesIO()
        font.save(buffer)
        return buffer.getvalue()
    except Exception:
        return None


def face(source: bytes | None, size: int, weight: int = 400) -> Font:
    """A face at `size`, instanced at `weight`, falling back to the font Pillow bundles."""
    if source is not None:
        try:
            font = ImageFont.truetype(io.BytesIO(source), size)
            font.set_variation_by_axes([weight])
            return font
        except OSError:
            pass

    return ImageFont.load_default(size=size)


def text_width(content: str, font: Font, tracking: int = 0) -> float:
    """The advance width of `content` in `font`, with `tracking` between glyphs."""
    if not content:
        return 0.0
    if tracking:
        return sum(font.getlength(char) for char in content) + tracking * (len(content) - 1)

    return font.getlength(content)


def shorten(content: str, max_width: float, font: Font, tracking: int = 0) -> str:
    """`content` cut back to a trailing ellipsis when it does not fit `max_width`."""
    if text_width(content, font, tracking) <= max_width:
        return content

    clipped = content
    while len(clipped) > 1 and text_width(clipped + ELLIPSIS, font, tracking) > max_width:
        clipped = clipped[:-1]

    return clipped.rstrip() + ELLIPSIS


def fit(content: str, max_width: float, sizes: tuple[int, int], load: Callable[..., Font]):
    """The text and the face to draw so a line fills `max_width` without passing it."""
    largest, smallest = sizes

    for size in range(largest, smallest, -1):
        font = load(size)
        if text_width(content, font) <= max_width:
            return content, font

    font = load(smallest)
    return shorten(content, max_width, font), font


def skills_line(skills: Sequence[str], max_width: float, font: Font) -> str:
    """As many leading skills as fit the width, or an empty string when none do."""
    for count in range(min(SKILLS_SHOWN, len(skills)), 0, -1):
        line = SKILLS_SEPARATOR.join(skills[:count])
        if text_width(line, font) <= max_width:
            return line

    return ""


def draw_tracked(
    draw: ImageDraw.ImageDraw, left: int, baseline: int, content: str, font: Font, fill: str
) -> None:
    """Draws `content` one glyph at a time so `DOMAIN_TRACKING` sits between them."""
    for index, char in enumerate(content):
        if index:
            left += DOMAIN_TRACKING
        draw.text((left, baseline), char, font=font, fill=fill, anchor="ls")
        left += font.getlength(char)


def circular_photo(photo: bytes) -> Image.Image | None:
    """The photo cropped to a circle of the card's diameter, inside the accent ring.

    Anything that is not a decodable image reads as no photo at all.
    """
    try:
        source = Image.open(io.BytesIO(photo))
        source.load()
        circle = ImageOps.fit(
            source.convert("RGB"), (PHOTO_DIAMETER, PHOTO_DIAMETER), Image.LANCZOS
        ).convert("RGBA")
    except Exception:
        return None

    size = PHOTO_DIAMETER * SUPERSAMPLE
    inset = RING_WIDTH * SUPERSAMPLE // 2
    mask = Image.new("L", (size, size), 0)
    ImageDraw.Draw(mask).ellipse((0, 0, size - 1, size - 1), fill=255)
    ring = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    ImageDraw.Draw(ring).ellipse(
        (inset, inset, size - 1 - inset, size - 1 - inset),
        outline=COLORS["accent"],
        width=RING_WIDTH * SUPERSAMPLE,
    )

    scaled = (PHOTO_DIAMETER, PHOTO_DIAMETER)
    circle.putalpha(mask.resize(scaled, Image.LANCZOS))
    circle.alpha_composite(ring.resize(scaled, Image.LANCZOS))
    return circle


def render_card(
    name: str, role: str, skills: Sequence[str], host: str, photo: bytes | None, assets_dir: Path
) -> bytes:
    """Renders the card as a PNG, in the faces `assets_dir` serves."""
    display = partial(face, truetype_bytes(assets_dir / DISPLAY_FONT_FILE), weight=700)
    body = partial(face, truetype_bytes(assets_dir / BODY_FONT_FILE))
    circle = circular_photo(photo) if photo else None

    card = Image.new("RGBA", (WIDTH, HEIGHT), COLORS["background"])
    if circle is not None:
        card.alpha_composite(circle, (PHOTO_LEFT, PHOTO_TOP))

    left = TEXT_LEFT_WITH_PHOTO if circle is not None else TEXT_LEFT_ALONE
    width = TEXT_RIGHT - left
    top = NAME_BASELINE_WITH_PHOTO if circle is not None else NAME_BASELINE_ALONE
    draw = ImageDraw.Draw(card)

    name_text, name_font = fit(name, width, NAME_SIZES, display)
    draw.text((left, top), name_text, font=name_font, fill=COLORS["name"], anchor="ls")

    role_text, role_font = fit(role, width, ROLE_SIZES, body)
    draw.text(
        (left, top + ROLE_OFFSET), role_text, font=role_font, fill=COLORS["role"], anchor="ls"
    )

    skills_font = body(SKILLS_SIZE, weight=600)
    line = skills_line(skills, width, skills_font)
    if line:
        draw.text(
            (left, top + SKILLS_OFFSET), line, font=skills_font, fill=COLORS["accent"], anchor="ls"
        )

    domain_font = body(DOMAIN_SIZE, weight=500)
    domain = shorten(host.upper(), width, domain_font, DOMAIN_TRACKING)
    draw_tracked(draw, left, DOMAIN_BASELINE, domain, domain_font, COLORS["domain"])

    output = io.BytesIO()
    card.convert("RGB").save(output, format="PNG", optimize=True)
    return output.getvalue()
